lib/mllp/client.ex

```diff
@@ -464,6 +470,51 @@ defmodule MLLP.Client do
     stop_connection(state, reason, "process terminated")
   end
 
+  defp recv_ack(_state, 0, _), do: {:error, :timeout}
+
+  defp recv_ack(state, timeout, buffer) do
+    {res, elapsed} = do_recv(state, 0, timeout)
+
+    case res do
+      {:ok, reply} ->
+        new_buf = buffer <> reply
+        check = byte_size(new_buf) - 3
+
+        case new_buf do
+          <<@header, _ack::binary-size(check), @trailer>> ->
+            {:ok, new_buf}
+
+          <<@header, _rest::binary>> ->
+            recv_ack(state, recv_timeout(timeout, elapsed), new_buf)
+
+          _ ->
+            {:error, :invalid_reply}
+        end
+
+      {:error, _} = err ->
+        err
+    end
+  end
+
+  defp do_recv(state, length, :infinity) do
+    res = state.tcp.recv(state.socket, length, :infinity)
+    {res, nil}
+  end
+
+  defp do_recv(state, length, timeout) do
+    t1 = System.convert_time_unit(:erlang.monotonic_time(), :native, :microsecond)
```

Comment:
I think this conversion may be pointless... depends on how you look at it. With the conversion from and to milliseconds, we lose precision. That is to say, as long as some milliseconds were spent in the recv op, then we will decrement the timeout by 1ms. Given that, we may as well simply measure in millisecond and if the elapsed time is 0, then we simply decrement the timeout by 1. 

The exception would be if 0 microseconds were spent (i.e., we spent nanoseconds in the recv op) then we don't decrement (i.e., the way it currently is).